aggregator/elem.go

```diff
@@ -0,0 +1,508 @@
+		values:   make([]timedGauge, 0, 2), // in most cases values will have two entries
+	}
+}
+
+// AddMetric adds a new gauge value
+func (e *GaugeElem) AddMetric(timestamp time.Time, mu unaggregated.MetricUnion) error {
+	alignedStart := timestamp.Truncate(e.policy.Resolution.Window).UnixNano()
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return errGaugeElemClosed
+	}
+	idx := e.findOrInsertWithLock(alignedStart)
+	e.values[idx].gauge.Add(mu.GaugeVal)
+	e.Unlock()
+	return nil
+}
+
+// ReadAndDiscard processes values before a given time and discards
+// them afterwards, returning whether the element can be collected
+// after discarding the values
+func (e *GaugeElem) ReadAndDiscard(earlierThan time.Time, fn aggMetricFn) bool {
+	earlierThanNs := earlierThan.UnixNano()
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return false
+	}
+	idx := 0
+	for range e.values {
+		// Bail as soon as the timestamp is no later than the target time
+		if e.values[idx].timeNs >= earlierThanNs {
+			break
+		}
+		endAt := time.Unix(0, e.values[idx].timeNs).Add(e.policy.Resolution.Window)
+		e.processValue(endAt, e.values[idx].gauge, fn)
+		idx++
+	}
+	if idx > 0 {
+		// Shift remaining values to the left and shrink the values slice
+		n := copy(e.values[0:], e.values[idx:])
+		e.values = e.values[:n]
+	}
+	canCollect := len(e.values) == 0 && e.tombstoned
+	e.Unlock()
+	return canCollect
+}
+
+// Close closes the gauge element
+func (e *GaugeElem) Close() {
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return
+	}
+	e.closed = true
+	e.id = nil
+	e.values = e.values[:0]
+	pool := e.opts.GaugeElemPool()
+	e.Unlock()
+
+	pool.Put(e)
+}
+
+// findOrInsertWithLock finds the element index whose timestamp matches
+// the start time passed in, or inserts one if it doesn't exist while
+// maintaining the time ascending order of the values. The returned index
+// is always valid.
+func (e *GaugeElem) findOrInsertWithLock(alignedStart int64) int {
+	numValues := len(e.values)
+	// Optimize for the common case
+	if numValues > 0 && e.values[numValues-1].timeNs == alignedStart {
+		return numValues - 1
+	}
+	// Binary search for the unusual case. We intentionally do not
+	// use the sort.Search() function because it requires passing
+	// in a closure
+	left, right := 0, numValues
+	for left < right {
+		mid := left + (right-left)/2 // avoid overflow
+		if e.values[mid].timeNs < alignedStart {
+			left = mid + 1
+		} else {
+			right = mid
+		}
+	}
+	// If the current timestamp is equal to or larger than the target time,
+	// return the index as is
+	if left < numValues && e.values[left].timeNs == alignedStart {
+		return left
+	}
+	// Otherwise we need to insert a new item
+	e.values = append(e.values, emptyTimedGauge)
+	// NB(xichen): it is ok if the source and the destination overlap
+	copy(e.values[left+1:numValues+1], e.values[left:numValues])
+	e.values[left] = timedGauge{
+		timeNs: alignedStart,
+		gauge:  aggregation.NewGauge(),
+	}
```

Comment:
Really kind of sucks we have to have this bucket searching implemented 3 times, there really is no way to generalize it or use a struct that is a union of the three types and we just use different field based on usecase?